dandi/files/__init__.py

```diff
@@ -0,0 +1,288 @@
+                    assert isinstance(bids2, BIDSDatasetDescriptionAsset)
+                    bidsdd = bids2
+                path_queue.extend((q, bidsdd) for q in p.iterdir())
+            elif any(p.iterdir()):
+                try:
+                    df = dandi_file(p, dandiset_path, bids_dataset_description=bidsdd)
+                except UnknownAssetError:
+                    if (p / BIDS_DATASET_DESCRIPTION).exists():
+                        bids2 = dandi_file(p / BIDS_DATASET_DESCRIPTION, dandiset_path)
+                        assert isinstance(bids2, BIDSDatasetDescriptionAsset)
+                        bidsdd = bids2
+                    path_queue.extend((q, bidsdd) for q in p.iterdir())
+                else:
+                    yield df
+        else:
+            df = dandi_file(p, dandiset_path, bids_dataset_description=bidsdd)
+            # Don't use isinstance() here, as GenericBIDSAsset's should still
+            # be returned
+            if type(df) is GenericAsset and not allow_all:
+                pass
+            elif isinstance(df, DandisetMetadataFile) and not (
+                allow_all or include_metadata
+            ):
+                pass
+            else:
+                yield df
+
+
+def dandi_file(
+    filepath: str | Path,
+    dandiset_path: Optional[str | Path] = None,
+    bids_dataset_description: Optional[BIDSDatasetDescriptionAsset] = None,
+) -> DandiFile:
+    """
+    Return a `DandiFile` instance of the appropriate type for the file at
+    ``filepath`` inside the Dandiset rooted at ``dandiset_path``.  If
+    ``dandiset_path`` is not set, it will default to ``filepath``'s parent
+    directory.
+
+    If ``bids_dataset_description`` is set, the file will be assumed to lie
+    within the BIDS dataset with the given :file:`dataset_description.json`
+    file at its root, resulting in a `BIDSAsset`.
+
+    If ``filepath`` is a directory, it must be of a type represented by a
+    `LocalDirectoryAsset` subclass; otherwise, an `UnknownAssetError` exception
+    will be raised.
+
+    A regular file named :file:`dandiset.yaml` will only be represented by a
+    `DandisetMetadataFile` instance if it is at the root of the Dandiset.
+
+    A regular file that is not of a known type will be represented by a
+    `GenericAsset` instance.
+    """
+    filepath = Path(filepath)
+    if dandiset_path is not None:
+        path = filepath.relative_to(dandiset_path).as_posix()
+        if path == ".":
+            raise ValueError("Dandi file path cannot equal Dandiset path")
+    else:
+        path = filepath.name
+    if filepath.is_file() and path == dandiset_metadata_file:
+        return DandisetMetadataFile(filepath=filepath)
+    if bids_dataset_description is None:
+        factory = DandiFileFactory()
+    else:
+        factory = BIDSFileFactory(bids_dataset_description)
+    return factory(filepath, path)
+
+
+class DandiFileType(Enum):
+    """:meta private:"""
```

Comment:
It keeps the classes from showing up in the Sphinx docs.  I don't want anyone other than dandi-cli developers to use these classes, and they can change or be removed without warning.

Reply:
Why not to prepend them with `_` to make them at least "protected".  I rarely read the documentation for every function/method/class to first figure out if it is public. I rely on Python convention of prefixes with `_` (protected) and `__` (private) for that.

Reply:
I don't think prefixing class names with `_` is standard practice.  Arguably, it violates PEP 8.

Reply:
hm indeed PEP 8 says "Class names should normally use the CapWords convention." for classes and "Use one leading underscore only for non-public methods and instance variables." only for methods/instance variables.  IMHO inconsistent as class definitions are pretty much "attributes" of a module here...  Some kind of a compromise could be to place such "non-public" elements into e.g. `/_private.py` and do not include them in `__all__` of the `__init__.py` so they would not be imported within `import *`. Moreover if someone would get such a type, it would be explicit from its path (not only from docstring which is generally not seen either).  WDYT?

Reply:
Private classes moved to `_private.py`.